### PR TITLE
Query namespaced resources in current namespace by default

### LINKED ISSUE #18: Query resources in current namespace by default
This came up in #9, #14 and #16.

Right now (v0.3.x) we are querying ALL objects in ALL api groups (without any namespace limitation). This has several problems:

- fails to query api X, because user has access to X in namespace1, but not in namespace2
- takes too long to enumerate all APIs across all namespaces (#16) 

From controller examples like Kubernetes Deployment, Knative, Agones, etc. it's fair to assume that **resources don't often create child resources in other namespaces**. 

### Proposed solution

When `kubectl tree` is querying child objects of a specified object:

- we should query only "namespaced" objects in the **same namespace as the specified object** as the _new default_
- we should offer `--all-namespaces` option to retain the current behavior (e.g. query all objects in all API groups –both namespaced and non-namespaced)

This addresses:

- slowness when used with large clusters with too many objects
- allow an escape hatch for the current behavior to still exist
- allow plugin to work when user has (full-) access to only some namespaces

It would likely work by refactorying query method to be namespace-scoped (query only namespaced APIs in the specified namespace) OR global-scoped (query both namespaced & non-namespaced APIs without specifying a namespace limitation, i.e. what we do today).

### PR DESCRIPTION
**Pull request by feloy**
Fixes #18

### CODE REVIEW

File: cmd/kubectl-tree/apis.go [hunk 1]
```diff
@@ -67,7 +67,9 @@ func findAPIs(client discovery.DiscoveryInterface) (*resourceMap, error) {
 				klog.V(4).Infof("    api (%s) doesn't have required verb, skipping: %v", apiRes.Name, apiRes.Verbs)
 				continue
 			}
-
+			if !*allNamespaces && !apiRes.Namespaced {
```
Comment (ahmetb) [inline cmd/kubectl-tree/apis.go]: Lets add a log.v(4) for skip reason

File: cmd/kubectl-tree/namespace.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package main
+
+var namespace *string
```
Comment (ahmetb) [inline cmd/kubectl-tree/namespace.go]: Lets not do global variables if we can :)

Comment (ahmetb) [reply cmd/kubectl-tree/namespace.go]: In this case we can we can just call the method twice, should be fairly simple. But lets move the var to the method scope.

File: cmd/kubectl-tree/rootcmd.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ import (
 
 var cf *genericclioptions.ConfigFlags
 
+var allNamespaces *bool
```
Comment (ahmetb) [inline cmd/kubectl-tree/rootcmd.go]: Avoid global var. 
pass as parameter to funcs.

File: README.md [hunk 4]
```diff
@@ -24,6 +24,12 @@ Example (Agones Fleet):
 
 ![example Kubernetes object hierarchy with Agones Fleet](assets/example-3.png)
 
+## Flags
+
+By default, the plugin will only search "namespaced" objects in the same namespace as the specified object.
```
Comment (ahmetb) [inline README.md]: FYI 80-char wrapped markdowns are good for you :)

You can install Markdownlint to your editor.

File: cmd/kubectl-tree/apis.go [hunk 5]
```diff
@@ -67,7 +67,10 @@ func findAPIs(client discovery.DiscoveryInterface) (*resourceMap, error) {
 				klog.V(4).Infof("    api (%s) doesn't have required verb, skipping: %v", apiRes.Name, apiRes.Verbs)
 				continue
 			}
-
+			if !allNs && !apiRes.Namespaced {
+				klog.V(4).Infof("    api (%s) is non-namespaced, skipping", apiRes.Name)
+				continue
+			}
```
Comment (ahmetb) [inline cmd/kubectl-tree/apis.go]: do you mind moving this skipping logic to to `getAllResources()`?
that's a more centralized place where we decide on what to do. This method just finds available APIs.

File: cmd/kubectl-tree/namespace.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package main
+
+func getNamespace() string {
+	ns := *cf.Namespace
+	if ns == "" {
```
Comment (ahmetb) [inline cmd/kubectl-tree/namespace.go]: while we're at it, let's shorten this by doing

```
if v: =*cf.Namespace; v != "" { return v }
rest of the method
```

File: cmd/kubectl-tree/rootcmd.go [hunk 7]
```diff
@@ -88,15 +98,7 @@ func run(_ *cobra.Command, args []string) error {
...
-	ns := *cf.Namespace
-	if ns == "" {
-		clientConfig := cf.ToRawKubeConfigLoader()
-		defaultNamespace, _, err := clientConfig.Namespace()
-		if err != nil {
-			defaultNamespace = "default"
-		}
-		ns = defaultNamespace
-	}
+	ns := getNamespace()
```
Comment (ahmetb) [inline cmd/kubectl-tree/rootcmd.go]: do you mind adding a log.v(2): namespace=%s allNamespaces=%v

so we can trace the runtime by looking at logs?

**Comment by ahmetb** [comment]
Thanks much.
This is good overall. Just added some code cleanup comments.
Then we're good.

File: README.md (hunk 4 above)
Comment (feloy) [reply README.md]: Useful tip, thanks

**Comment by feloy** [comment]
I think I've made the changes.

File: cmd/kubectl-tree/rootcmd.go [hunk 8]
```diff
@@ -134,6 +137,9 @@ func init() {
 	})
 
 	cf = genericclioptions.NewConfigFlags(true)
+
+	rootCmd.Flags().BoolP(allNamespacesFlag, "A", false, "query all objects in all API groups, both namespaced and non-namespaced")
```
Comment (ahmetb) [inline cmd/kubectl-tree/rootcmd.go]: I'll merge this but let's make another PR to simplify this like:

> query all namespaces for child objects

File: README.md [hunk 9]
```diff
@@ -24,6 +24,14 @@ Example (Agones Fleet):
 
 ![example Kubernetes object hierarchy with Agones Fleet](assets/example-3.png)
 
+## Flags
```
Comment (ahmetb) [inline README.md]: IMO this doesn't deserve a readme section like this, feel free to remove in a follow-up PR.

The behavior is pretty intuitive, I don't even expect people to note the difference. It's a very rare case someone will run into this.

**Pull request merged**